File: packages/api/src/models/ApiKey.ts
/**
 * ApiKey Model — Personal access tokens for programmatic API access
 */

import { DataTypes, Model, Optional } from 'sequelize';
import { sequelize } from '../db';
import crypto from 'node:crypto';

interface ApiKeyAttributes {
  id: string;
  userId: string;
  name: string;
  keyHash: string;
  keyPrefix: string;
  lastUsedAt: Date | null;
  createdAt: Date;
  updatedAt: Date;
}

interface ApiKeyCreationAttributes extends Optional<ApiKeyAttributes, 'id' | 'lastUsedAt' | 'createdAt' | 'updatedAt'> {}

export class ApiKey extends Model<ApiKeyAttributes, ApiKeyCreationAttributes> implements ApiKeyAttributes {
  public id!: string;
  public userId!: string;
  public name!: string;
  public keyHash!: string;
  public keyPrefix!: string;
  public lastUsedAt!: Date | null;
  public readonly createdAt!: Date;
  public readonly updatedAt!: Date;
}

const API_KEY_PREFIX = 'rpk_';

export function generateApiKey(): { plainKey: string; keyHash: string; keyPrefix: string } {
  const raw = crypto.randomBytes(32).toString('hex');
  const plainKey = `${API_KEY_PREFIX}${raw}`;
  const keyHash = crypto.createHash('sha256').update(plainKey).digest('hex');
  const keyPrefix = plainKey.substring(0, 10); // "rpk_" + first 6 chars
  return { plainKey, keyHash, keyPrefix };
}

export function hashApiKey(plainKey: string): string {
  return crypto.createHash('sha256').update(plainKey).digest('hex');
}

export function isApiKeyFormat(token: string): boolean {
  return token.startsWith(API_KEY_PREFIX);
}

ApiKey.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    userId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: { model: 'users', key: 'id' },
      onDelete: 'CASCADE',
    },
    name: {
      type: DataTypes.STRING(100),
      allowNull: false,
      validate: { len: [1, 100] },
    },
    keyHash: {
      type: DataTypes.STRING(64),
      allowNull: false,
    },
    keyPrefix: {
      type: DataTypes.STRING(10),
      allowNull: false,
    },
    lastUsedAt: {
      type: DataTypes.DATE,
      allowNull: true,
    },
    createdAt: { type: DataTypes.DATE, allowNull: false },
    updatedAt: { type: DataTypes.DATE, allowNull: false },
  },
  {
    sequelize,
    tableName: 'api_keys',
    timestamps: true,
    indexes: [
      { fields: ['user_id'] },
      { fields: ['key_hash'], unique: true },
      { fields: ['key_prefix'] },
    ],
  }
);

File: packages/server/app/models/api_key.py
"""ApiKey model — mirrors the Node.js Sequelize ApiKey model exactly."""

import hashlib
import os
import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, String, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db import Base

API_KEY_PREFIX = 'rpk_'


def generate_api_key() -> tuple[str, str, str]:
    """Generate a new API key tuple: (plain_key, key_hash, key_prefix)."""
    raw = os.urandom(32).hex()
    plain_key = f'{API_KEY_PREFIX}{raw}'
    key_hash = hashlib.sha256(plain_key.encode()).hexdigest()
    key_prefix = plain_key[:10]  # "rpk_" + first 6 chars
    return plain_key, key_hash, key_prefix


def hash_api_key(plain_key: str) -> str:
    """SHA-256 hash an API key for lookup."""
    return hashlib.sha256(plain_key.encode()).hexdigest()


def is_api_key_format(token: str) -> bool:
    """Check if a token string looks like an API key."""
    return token.startswith(API_KEY_PREFIX)


class ApiKey(Base):
    """Personal access token for programmatic API access."""

    __tablename__ = 'api_keys'

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        'user_id',
        UUID(as_uuid=True),
        ForeignKey('users.id'),
        nullable=False,
        index=True,
    )

    user: Mapped['User'] = relationship('User', back_populates='api_keys')
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    key_hash: Mapped[str] = mapped_column(
        'key_hash',
        String(64),
        unique=True,
        nullable=False,
    )
    key_prefix: Mapped[str] = mapped_column(
        'key_prefix',
        String(10),
        nullable=False,
        index=True,
    )
    last_used_at: Mapped[datetime | None] = mapped_column(
        'last_used_at',
        DateTime(timezone=True),
        nullable=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        'created_at',
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        'updated_at',
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )
